File: src/strings/valid_parenthesis.py
from collections import deque


class Solution:
    def isValid(self, string: str) -> bool:
        closeOpenMap = {"}": "{", "]": "[", ")": "("}
        stack = deque()
        for char in string:
            if char in closeOpenMap:
                if stack and stack[-1] == closeOpenMap[char]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(char)
        return True if not stack else False

File: java/src/main/java/com/blind75/strings/ValidParenthesis.java
package com.blind75.strings;

import java.util.ArrayList;
import java.util.Map;
import java.util.Stack;

public class ValidParenthesis {

    private Map<Character, Character> parenthesisMap = Map.of(
            ')', '(',
            '}', '{',
            ']', '['
    );

    public boolean isValid(String inputString) {
        var stack = new Stack<Character>();
        var openingParenthesis = new ArrayList<Character>(parenthesisMap.values());
        var closingParenthesis = new ArrayList<Character>(parenthesisMap.keySet());

        for (var characterByte : inputString.getBytes()) {
            var character = Character.valueOf((char) characterByte);
            if (openingParenthesis.contains(character)) {
                stack.push(character);
            } else if (closingParenthesis.contains(character)) {
                if (stack.isEmpty()) {
                    return false;
                }
                var charFromStack = stack.pop();
                if (!charFromStack.equals(parenthesisMap.get(character)))
                    return false;
            }
        }
        if (stack.empty()) {
            return true;
        }
        return false;
    }
}
